crates/bevy_math/src/primitives/dim3.rs

```diff
@@ -359,22 +359,25 @@ impl Primitive3d for Line3d {}
     reflect(Serialize, Deserialize)
 )]
 pub struct Segment3d {
-    /// The direction of the line
-    pub direction: Dir3,
-    /// Half the length of the line segment. The segment extends by this amount in both
-    /// the given direction and its opposite direction
-    pub half_length: f32,
+    /// The first point of the segment
+    pub point1: Vec3,
+    /// The second point of the segment
+    pub point2: Vec3,
 }
 impl Primitive3d for Segment3d {}
 
 impl Segment3d {
+    /// Create a new `Segment3d` from it's points
+    #[inline(always)]
+    pub fn new(point1: Vec3, point2: Vec3) -> Self {
+        Self { point1, point2 }
+    }
+
     /// Create a new `Segment3d` from a direction and full length of the segment
     #[inline(always)]
-    pub fn new(direction: Dir3, length: f32) -> Self {
-        Self {
-            direction,
-            half_length: length / 2.0,
-        }
+    pub fn from_direction_and_length(direction: Dir3, length: f32) -> Self {
+        let half_length = length / 2.;
+        Self::new(direction * -half_length, direction * -half_length)
```

Comment:
This produces a segment with zero length.
```suggestion
    pub fn from_direction_and_length(direction: Dir3, length: f32) -> Self {
        let half_length = length / 2.;
        Self::new(direction * -half_length, direction * half_length)
```